grandpa/verifier/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+    ParachainHeaderProofs, ParachainHeadersWithFinalityProof,
+};
+use sp_core::H256;
+use sp_runtime::traits::Header;
+use sp_trie::{LayoutV0, StorageProof};
+
+/// This function verifies the GRANDPA finality proof for both standalone chain and parachain
+/// headers.
+///
+/// TODO: return verified header and the associated time stamp
+pub fn verify_grandpa_finality_proof<H, Host>(
+    mut consensus_state: ConsensusState,
+    finality_proof: FinalityProof<H>,
+) -> Result<(ConsensusState, Header), error::Error>
+where
+    H: Header<Hash = H256, Number = u32>,
+    H::Number: finality_grandpa::BlockNumberOps + Into<u32>,
+    Host: HostFunctions,
+    Host::BlakeTwo256: Hasher<Out = H256>,
+{
+    // First validate unknown headers.
+    let headers = AncestryChain::<H>::new(&finality_proof.unknown_headers);
+
+    let target = finality_proof
+        .unknown_headers
+        .iter()
+        .max_by_key(|h| *h.number())
+        .ok_or_else(|| anyhow!("Unknown headers can't be empty!"))?;
+
+    // this is illegal
+    if target.hash() != finality_proof.block {
+        Err(anyhow!("Latest finalized block should be highest block in unknown_headers"))?;
+    }
+
+    let justification = GrandpaJustification::<H>::decode(&mut &finality_proof.justification[..])?;
+
+    if justification.commit.target_hash != finality_proof.block {
+        Err(anyhow!("Justification target hash and finality proof block hash mismatch"))?;
+    }
+
+    let from = consensus_state.latest_hash;
+
+    let base = finality_proof
+        .unknown_headers
+        .iter()
+        .min_by_key(|h| *h.number())
+        .ok_or_else(|| anyhow!("Unknown headers can't be empty!"))?;
+
+    if base.number() < &consensus_state.latest_height {
+        headers.ancestry(base.hash(), consensus_state.latest_hash).map_err(|_| {
+            anyhow!(
+                "[verify_grandpa_finality_proof] Invalid ancestry (base -> latest relay block)!"
+            )
+        })?;
+    }
+
+    let mut finalized = headers
+        .ancestry(from, target.hash())
+        .map_err(|_| anyhow!("[verify_grandpa_finality_proof] Invalid ancestry!"))?;
+    finalized.sort();
+
+    // 2. verify justification.
+    justification
+        .verify::<Host>(consensus_state.current_set_id, &consensus_state.current_authorities)?;
+
+    Ok((consensus_state, base))
+}
+/// This function verifies the GRANDPA finality proof for relay chain headers.
+///
+/// Next, we prove the finality of parachain headers, by verifying patricia-merkle trie state proofs
+/// of these headers, stored at the recently finalized relay chain heights.
+/// TODO: return verified headers(compared with the consensus state para ids) and the associated
+/// time stamp TODO: remove all host functions
+pub fn verify_parachain_headers_with_grandpa_finality_proof<H, Host>(
+    mut consensus_state: ConsensusState,
+    proof: ParachainHeadersWithFinalityProof<H>,
+) -> Result<(ConsensusState, Vec<Header>), error::Error>
```

Comment:
Return  a map of para ids to a tuple of parachain header and timestamp, extract timestamp from the extrinsic.